src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -0,0 +1,532 @@
+      help="Arguments to pass directly to setup.py, e.g. "
+           "`--setup-py2-args=\"bdist_wheel\"`. If unspecified, we just dump the setup.py chroot."
+    )
+
+
+class SetupPy(Goal):
+  """Runs setup.py commands."""
+  subsystem_cls = SetupPyOptions
+
+
+@console_rule
+async def run_setup_pys(addresses: BuildFileAddresses, console: Console, options: SetupPyOptions,
+                        distdir: DistDir, workspace: Workspace) -> SetupPy:
+  """Run setup.py commands on all exported targets addressed."""
+  targets = await Get[HydratedTargets](BuildFileAddresses, addresses)
+
+  exported_targets: List[ExportedTarget] = []
+  nonexported_targets: List[HydratedTarget] = []
+  for target in targets:
+    if _is_exported(target):
+      exported_targets.append(ExportedTarget(target))
+    else:
+      nonexported_targets.append(target)
+  if nonexported_targets:
+    raise TargetNotExported(
+      'Cannot run setup.py on these targets, because they have no `provides=` clause: '
+      f'{", ".join(so.address.reference() for so in nonexported_targets)}')
+
+  for target in exported_targets:
+    addr = target.hydrated_target.address.reference()
+    chroot = await Get[SetupPyChroot](SetupPyChrootRequest(target))
+    provides = target.hydrated_target.adaptor.provides
+    setup_py_dir = distdir.relpath / f'{provides.name}-{provides.version}'
+    if options.values.args:
+      console.print_stderr(f'Running setup.py for {addr}')
+      setup_py_result = await Get[RunSetupPyResult](
+        RunSetupPyRequest(target, chroot, tuple(options.values.args)))
+      result_digest = setup_py_result.output
+    else:
+      # Just dump the chroot.
+      result_digest = chroot.digest
+    workspace.materialize_directory(
+      DirectoryToMaterialize(result_digest, path_prefix=str(setup_py_dir))
+    )
+    console.print_stderr(f'Wrote setup.py chroot for {addr} to {setup_py_dir}')
+
+  return SetupPy(0)
+
+
+# We write .py sources into the chroot under this dir.
+CHROOT_SOURCE_ROOT = 'src'
+
+
+SETUP_BOILERPLATE = """
+# DO NOT EDIT THIS FILE -- AUTOGENERATED BY PANTS
+# Target: {target_address_spec}
+
+from setuptools import setup
+
+setup(**{setup_kwargs_str})
+"""
+
+
+@rule
+async def run_setup_py(
+    req: RunSetupPyRequest,
+    setuptools_setup: SetuptoolsSetup,
+    python_setup: PythonSetup,
+    subprocess_encoding_environment: SubprocessEncodingEnvironment
+) -> RunSetupPyResult:
+  """Run a setup.py command on a single exported target."""
+  merged_input_files = await Get[Digest](
+    DirectoriesToMerge(directories=(
+      req.chroot.digest,
+      setuptools_setup.requirements_pex.directory_digest))
+  )
+  request = setuptools_setup.requirements_pex.create_execute_request(
+    python_setup=python_setup,
+    subprocess_encoding_environment=subprocess_encoding_environment,
+    pex_path="./setuptools.pex",
+    pex_args=('setup.py', *req.args),
+    input_files=merged_input_files,
+    output_files=('**',),
```

Comment:
Hm... globs are not supported by the remote execution API. This works because we use our glob expansion code to capture outputs locally, but we should probably make globs an error in `ExecuteProcessRequest`.

But also: you probably don't want to actually capture _everything_ here, do you? You'll capture the pex you were running, for example. Consider writing to an output directory and then capturing that?

Reply:
Good to know. So if I don't know the names of output files in advance I use `output_directories`? 

You're right that I don't want to capture everything. But I'm not sure what the correct surface area is.  All dists created by running`setup.py` commands should be under `dist/` in the chroot, so maybe that's enough for now.